libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp

```diff
@@ -0,0 +1,116 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <algorithm>
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+// template<input_iterator I, sentinel_for<I> S, class T,
+//          indirectly-binary-left-foldable<T, I> F>
+//   constexpr see below ranges::fold_left_with_iter(I first, S last, T init, F f);
+//
+// template<input_range R, class T, indirectly-binary-left-foldable<T, iterator_t<R>> F>
+//   constexpr see below ranges::fold_left_with_iter(R&& r, T init, F f);
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+#include <functional>
+
+#include "test_range.h"
+#include "../gaussian_sum.h"
+
+constexpr bool test() {
```

Comment:
I would add runtime tests for:

1. An empty range
2. A range where summing in any order other than `fold_left` would result in an overflow or other invalid condition.
3. That we don't copy the functor like we intend.
4. That the functor is applied the correct amount of times.
5. On non-integral or builtin types.
6. On types other than vector & its iterators.

I understand the signatures for a lot of these configurations are checked via meta-programming elsewhere, but we need to run code to test them too. What if there's a lifetime bug that wouldn't be caught unless the type were' folding allocates?

Reply:
* (1) is already done (lines 29--41)
* (2) and (5) can be coupled together using `std::basic_string<some-small-allocator>`.
* I'm not sure I understand what you're asking for in (3). I've already got cases where the function object is copied and when it's directly initialised.
* Thanks for catching (4)
* I'm not really sure we'll get more value out of (6) than `return_types.compile.pass.cpp`, which checks this.

Reply:
* Please add more tests for (1). What if we're dereferencing the first iterator? In all those cases that's valid because it's actually backed by an element.

(3) What's the purpose of wrapping the binary functor in `std::ref`? To prevent it from being copied repeatedly in case it has state. It would be a meaningful regression were we to start copying the binary operator more.

(6) Almost all of the tests are over contiguous ranges of densely packed objects. So code using more opaque or non-contiguous iterators will compile entirely differently, have wildly different memory access patterns, and could tickle all sorts of bugs we don't know we have.

Reply:
(1) Fair enough.
(3) Ah, good idea.
(6) I still don't think that's going to expose any issues in `fold_left`. I'll change to different types (it's hardly any effort), but still I'm not convinced we're going to get broader runtime coverage.

Reply:
Completed.